libcentrifugo/engineredis.go

```diff
@@ -613,28 +625,129 @@ func (e *RedisEngine) runPubSub() {
 	}
 }
 
-func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
+type pubRequest struct {
+	channel     ChannelID
+	message     []byte
+	messageJSON []byte
+	historyKey  string
+	opts        *publishOpts
+	err         *chan error
+}
+
+func (pr *pubRequest) done(err error) {
+	if pr.err == nil {
+		return
+	}
+	*(pr.err) <- err
+}
+
+func (pr *pubRequest) result() error {
+	if pr.err == nil {
+		// No waiting, as caller didn't care about response
+		return nil
+	}
+	return <-*(pr.err)
+}
+
+func fillPublishBatch(ch chan *pubRequest, prs *[]*pubRequest) {
+	for len(*prs) < RedisPublishBatchLimit {
```

Comment:
Doesn't respect `RedisPublishBatchLimit` explicitly. For now I *guess* that's not a big deal because that limit is larger than buffer size, but I don't know what guarantees Go spec makes about this.

If another thread is pushing new things in constantly and the underlying implementation is a ring-buffer, presumably we could go round the ring many times and consume more than buffer size messages... Seems very unlikely in practice, but if we have a limit constant we should probably respect it explicitly.